File: 백준/Silver/1629. 곱셈/곱셈.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split(" ");
let [a, b, c] = input.map(BigInt);

function modPow(base, exp, mod) {
  if (exp === 0n) return 1n;
  if (exp === 1n) return base % mod;

  let half = modPow(base, exp / 2n, mod);
  let result = (half * half) % mod;

  if (exp % 2n === 1n) {
    result = (result * base) % mod;
  }
  return result;
}

console.log(modPow(a, b, c).toString());

File: 백준/Silver/1629. 곱셈/곱셈.py
a,b,c = map(int, input().split())


def recur(base,exp,mod):
  if exp == 0:
    return 1
  if exp == 1:
    return base % mod
  
  # 여기서 
  half = recur(base, exp // 2, mod)
  result = (half * half) % mod

  if exp % 2 == 1:
    result = (result * base) % mod
  return result

print(recur(a,b,c))